efficientdet/keras/train.py

```diff
@@ -0,0 +1,266 @@
+    for level in range(config.get('min_level'), config.get('max_level') + 1):
+
+      def _can_partition(spatial_dim):
+        partitionable_index = np.where(
+            spatial_dim % np.array(FLAGS.input_partition_dims) == 0)
+        return len(partitionable_index[0]) == len(FLAGS.input_partition_dims)
+
+      spatial_dim = feat_sizes[level]
+      if _can_partition(spatial_dim['height']) and _can_partition(
+          spatial_dim['width']):
+        labels_partition_dims['box_targets_%d' %
+                              level] = FLAGS.input_partition_dims
+        labels_partition_dims['cls_targets_%d' %
+                              level] = FLAGS.input_partition_dims
+      else:
+        labels_partition_dims['box_targets_%d' % level] = None
+        labels_partition_dims['cls_targets_%d' % level] = None
+    num_cores_per_replica = FLAGS.num_cores_per_replica
+    input_partition_dims = [FLAGS.input_partition_dims, labels_partition_dims]
+    num_shards = FLAGS.num_cores // num_cores_per_replica
+  else:
+    num_cores_per_replica = None
+    input_partition_dims = None
+    num_shards = FLAGS.num_cores
+
+  params = dict(config.as_dict(),
+                model_name=FLAGS.model_name,
+                iterations_per_loop=FLAGS.iterations_per_loop,
+                model_dir=FLAGS.model_dir,
+                num_shards=num_shards,
+                num_examples_per_epoch=FLAGS.num_examples_per_epoch,
+                strategy=FLAGS.strategy,
+                backbone_ckpt=FLAGS.backbone_ckpt,
+                ckpt=FLAGS.ckpt,
+                val_json_file=FLAGS.val_json_file,
+                testdev_dir=FLAGS.testdev_dir,
+                mode=FLAGS.mode)
+
+  # set mixed precision policy by keras api which will generate mixed precision model
+  precision = utils.get_precision(params['strategy'], params['mixed_precision'])
+  policy = tf.keras.mixed_precision.Policy(precision)
+  tf.keras.mixed_precision.set_policy(policy)
+
+  def get_dataset(input_context, is_training, params):
+    # TODO(fsx950223): ensure it's global batch size or per batch size
+    batch_size = FLAGS.train_batch_size if is_training else FLAGS.eval_batch_size
+    file_pattern = FLAGS.training_file_pattern if is_training else FLAGS.validation_file_pattern
+    params = copy.deepcopy(params)
+    params.update(dict(batch_size=batch_size))
+    return dataloader.InputReader(
+        file_pattern,
+        is_training=is_training,
+        use_fake_data=FLAGS.use_fake_data,
+        max_instances_per_image=config.max_instances_per_image)(params)
+
+  if strategy:
+    train_dataset = strategy.experimental_distribute_datasets_from_function(
+        functools.partial(get_dataset, is_training=True, params=params))
+    eval_dataset = strategy.experimental_distribute_datasets_from_function(
+        functools.partial(get_dataset, is_training=False, params=params))
+
+  # TODO(fsx950223): add train logic
```

Comment:
After you adding the train logic, I would be happy to test it on TPU.

Reply:
Use custom loop or keras.model.fit?

Reply:
Either way is good to me. I guess custom loop is more flexible? WDYT?

Reply:
https://www.tensorflow.org/guide/keras/customizing_what_happens_in_fit
I prefer custom loop too before I read above tutorial.